nixos/modules/config/users-groups.nix

```diff
@@ -528,32 +563,117 @@ in {
       render.gid = ids.gids.render;
     };
 
-    system.activationScripts.users = stringAfter [ "stdio" ]
+    system.activationScripts.users-before-etc =
       ''
         install -m 0700 -d /root
-        install -m 0755 -d /home
+      '' + optionalString (!accountsAreDeterministic) ''
 
         ${pkgs.perl}/bin/perl -w \
           -I${pkgs.perlPackages.FileSlurp}/${pkgs.perl.libPrefix} \
           -I${pkgs.perlPackages.JSON}/${pkgs.perl.libPrefix} \
           ${./update-users-groups.pl} ${spec}
       '';
 
+    # Allow either "users-before-etc" or "etc" to create passwd/group; either
+    # way, "users" is late enough to create home directories. Note: in case the
+    # specified group doesn't exist, let chown look up what primary group the
+    # user actually got assigned.
+    system.activationScripts.users = stringAfter [ "etc" ] (
+      concatMapStrings (u: optionalString u.createHome ''
+        test -d ${u.home} || install -m 0700 -d ${u.home}
+        chown ${u.name}: ${u.home}
+      '') (attrValues cfg.users)
+    );
+
+    # As long as /etc/shadow is only readable by root, it's at least as secure
+    # as all the passwords which went into it, no matter how we construct it.
+    # TODO: make this a systemd service wantedBy/before nss-user-lookup.target
+    # because neither activation scripts nor systemd need to check passwords
+    # but it also needs to run during activation on an already-booted system
+    system.activationScripts.shadow = stringAfter [ "users" ] (optionalString accountsNeedShadow (
```

Comment:
Why after `users`?

Reply:
I don't remember, and re-reading the diffs isn't reminding me. Hmmm.